### PR TITLE
Migrate bokehjs' components to use shadow DOM

### LINKED ISSUE #9757: Investigate using shadow DOM to gain proper isolation
Relies on #9731 as shadow DOM is not supported in legacy browsers.

### PR DESCRIPTION
**Pull request by mattpap**
This supersedes PR #10297. More likely than not, this PR will only cover a few areas of bokehjs (so far toolbars and menus), and I will be working on the rest in subsequent PRs.

- [x] fixes #9757

### CODE REVIEW

**Comment by bryevdv** [comment]
Getting back in to things, I will take a look at this today and tomorrow

File: sphinx/source/docs/user_guide/extensions.rst [hunk 1]
```diff
@@ -107,7 +107,7 @@ the final BokehJS scripts.
         // slider value.
         super.render()
 
-        this.el.appendChild(div({
+        this.shadow_el.appendChild(div({
```
Comment (bryevdv) [inline sphinx/source/docs/user_guide/extensions.rst:110]: If there is nothing we can do to automatically ease this transition, we need to at least describe necessary changes in the migration guide

**Comment by bryevdv** [comment]
@mattpap FYI built locally on OSX it now seems any slider movement also selects the slider title text
Screen Shot 2021-12-29 at 16 40 05

**Comment by bryevdv** [comment]
@mattpap can you also summarize when the shadow DOM is/is not utilized? So far I only notice it employed in Bokeh server apps.

**Comment by mattpap** [comment]
It's utilized by all components (i.e. views that extend `DOMComponentView`), except for `models.dom`, which represent raw DOM elements (though in that case I will later add an option for transforming a DOM node into a component with shadow DOM attached). Note however that CSS modularization is not fully completed in this PR, e.g. toolbar and tool buttons share DOM and CSS, whereas they should be separate; similarly buttons and button groups, and dropdown menus need further refinement.

**Comment by mattpap** [comment]
> FYI built locally on OSX it now seems any slider movement also selects the slider title text

Some breakage involving external libraries is expected at this stage (note this is a transitional PR), as those are usually not well suited for CSS modularization. I already dealt with date picker, slick grid, choices, at al.

File: sphinx/source/docs/user_guide/extensions.rst (hunk 1 above)
Comment (mattpap) [reply sphinx/source/docs/user_guide/extensions.rst:110]: I'm not sure at this stage (as I said earlier, this is a transitional PR) if this API will be kept as-is, especially given the extent of changes to come. When the APIs settle, then we can think about migration guide.

File: bokehjs/src/less/widgets/choices.less [hunk 2]
```diff
@@ -1,65 +1,63 @@
...
+}
 
-  .choices[data-type*=select-multiple] .choices__button.solid {
-    background-image: url(data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjEiIGhlaWdodD0iMjEiIHZpZXdCb3g9IjAgMCAyMSAyMSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48ZyBmaWxsPSIjZmZmZmZmIiBmaWxsLXJ1bGU9ImV2ZW5vZGQiPjxwYXRoIGQ9Ik0yLjU5Mi4wNDRsMTguMzY0IDE4LjM2NC0yLjU0OCAyLjU0OEwuMDQ0IDIuNTkyeiIvPjxwYXRoIGQ9Ik0wIDE4LjM2NEwxOC4zNjQgMGwyLjU0OCAyLjU0OEwyLjU0OCAyMC45MTJ6Ii8+PC9nPjwvc3ZnPg==);
-    border-left: 1px solid white;
-    opacity: 1;
-  }
+.choices[data-type*=select-multiple] .choices__button.solid {
+  background-image: url(data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjEiIGhlaWdodD0iMjEiIHZpZXdCb3g9IjAgMCAyMSAyMSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48ZyBmaWxsPSIjZmZmZmZmIiBmaWxsLXJ1bGU9ImV2ZW5vZGQiPjxwYXRoIGQ9Ik0yLjU5Mi4wNDRsMTguMzY0IDE4LjM2NC0yLjU0OCAyLjU0OEwuMDQ0IDIuNTkyeiIvPjxwYXRoIGQ9Ik0wIDE4LjM2NEwxOC4zNjQgMGwyLjU0OCAyLjU0OEwyLjU0OCAyMC45MTJ6Ii8+PC9nPjwvc3ZnPg==);
```
[the remark this answers is not in the document]
Comment (mattpap) [reply bokehjs/src/less/widgets/choices.less:60]: It's supposed to be "cross" icon, at least according to variable name that holds it (in choices' scss).

[the remark this answers is not in the document]
Comment (bryevdv) [reply bokehjs/src/less/widgets/choices.less:60]: 🤷

File: bokeh/_testing/util/selenium.py [hunk 3]
```diff
@@ -81,13 +81,155 @@
...
+    elements = find_matching_elements(driver, selector, root=root)
+    n = len(elements)
+    if n == 0:
+        raise ValueError("not found")
+    else:
+        return elements[0]
+    #elif n == 1:
+    #    return elements[0]
+    #else:
+    #    raise ValueError("multiple elements found")
```
[the remark this answers is not in the document]
Comment (mattpap) [reply bokeh/_testing/util/selenium.py:124]: I think the commented code can be re-enabled now. `find_matching_elemetn()` is a measure of last resort anyway, and should be used with caution.

File: bokeh/core/_templates/notebook_load.html [hunk 4]
```diff
@@ -21,8 +21,16 @@
...
...iH3gmKoiJDU7QpLgoLjLIQCpEsNJ1vqUOdO7ppbuec5+V+rj4ctwzd8IIbbi6u+8f1539dt3A78eXC7QizUF7gyV1fD1Yqg4JWz84yffhm0qkFqBogB9rM8tZdtwVsPUhWhGcFJngGeWrPzHm5oaMmkfEg1usvLFyc8jLRqDOMru7AyC8saQr7GG7f5fvDeH7Ej8CM66nIF+8yngt6HWaKh7k49Soy9nXurCi1o3qUbS3zWfrYeQDTB/Qj6kX6Ybhw4B+bOYoLKCC9H3Nu/leUTZ1JdRWkkn2ldcCamzrcf47KKXdAJllSlxAOkRgyHsGC/zRday5Qld9DyoM4/q/rUoy/CXh3jzOu3bHUVZeU+DEn8FInkPBFlu3+nW3Nw0mk6vCDiWg8CeJaxEwuHS3+z5RgY+YBR6V1Z1nxSOfoaPa4LASWxxdNp+VWTk7+4vzaou8v8PN+xo+KY2xsw6une2frhw05CTYOmQvsEhjhWjn0bmXPjpE1+kplmmkP3suftwTubK9Vq22qKmrBhpY4jvd5afdRA3wGjFAgcnTK2s4hY0/GPNIb0nErGMCRxWOOX64Z8RAC4oCXdklmEvcL8o0BfkNK4lUg9HTl+oPlQxdNo3Mg4Nv175e/1LDGzZen30MEjRUtmXSfiTVu1kK8W4txyV6BMKlbgk3lMwYCiusNy9fVfvvwMxv8Ynl6vxoByANLTWplvuj/nF9m2+PDtt1eiHPBr1oIfhCChQMBw6Aw0UulqTKZdfVvfG7VcfIqLG9bcldL/+pdWTLxLUy8Qq38heUIjh4XlzZxzQm19lLFlr8vdQ97rjZVOLf8nclzckbcD4wxXMidpX30sFd37Fv/GtwwhzhxGVAprjbg0gCAEeIgwCZyTV2Z1REEW8O4py0wsjeloKoMr6iCY6dP92H6Vw/oTyICIthibxjm/DfN9lVz8IqtqKYLUXfoKVMVQVVJOElGjrnnUt9T9wbgp8AyYKaGlqingHZU/uG2NTZSVqwHQTWkx9hxjkpWDaCg6Ckj5qebgBVbT3V3NNXMSiWSDdGV3hrtzla7J+duwPOToIg42ChPQOQjspnSlp1V+Gjdged7+8UN5CRAV7a5EdFNwCjEaBR27b3W890TE7g24NAP/mMDXRWrGoFPQI9ls/MWO2dWFAar/xcOIImbbpA3zgAAAABJRU5ErkJggg==);
```
[the remark this answers is not in the document]
Comment (mattpap) [reply bokeh/core/_templates/notebook_load.html:29]: Possibly, though I will be later revisiting changes like this one. For now this is here to make things work.

File: bokehjs/src/less/logo.less [hunk 5]
```diff
@@ -1,24 +1,16 @@
...
...iH3gmKoiJDU7QpLgoLjLIQCpEsNJ1vqUOdO7ppbuec5+V+rj4ctwzd8IIbbi6u+8f1539dt3A78eXC7QizUF7gyV1fD1Yqg4JWz84yffhm0qkFqBogB9rM8tZdtwVsPUhWhGcFJngGeWrPzHm5oaMmkfEg1usvLFyc8jLRqDOMru7AyC8saQr7GG7f5fvDeH7Ej8CM66nIF+8yngt6HWaKh7k49Soy9nXurCi1o3qUbS3zWfrYeQDTB/Qj6kX6Ybhw4B+bOYoLKCC9H3Nu/leUTZ1JdRWkkn2ldcCamzrcf47KKXdAJllSlxAOkRgyHsGC/zRday5Qld9DyoM4/q/rUoy/CXh3jzOu3bHUVZeU+DEn8FInkPBFlu3+nW3Nw0mk6vCDiWg8CeJaxEwuHS3+z5RgY+YBR6V1Z1nxSOfoaPa4LASWxxdNp+VWTk7+4vzaou8v8PN+xo+KY2xsw6une2frhw05CTYOmQvsEhjhWjn0bmXPjpE1+kplmmkP3suftwTubK9Vq22qKmrBhpY4jvd5afdRA3wGjFAgcnTK2s4hY0/GPNIb0nErGMCRxWOOX64Z8RAC4oCXdklmEvcL8o0BfkNK4lUg9HTl+oPlQxdNo3Mg4Nv175e/1LDGzZen30MEjRUtmXSfiTVu1kK8W4txyV6BMKlbgk3lMwYCiusNy9fVfvvwMxv8Ynl6vxoByANLTWplvuj/nF9m2+PDtt1eiHPBr1oIfhCChQMBw6Aw0UulqTKZdfVvfG7VcfIqLG9bcldL/+pdWTLxLUy8Qq38heUIjh4XlzZxzQm19lLFlr8vdQ97rjZVOLf8nclzckbcD4wxXMidpX30sFd37Fv/GtwwhzhxGVAprjbg0gCAEeIgwCZyTV2Z1REEW8O4py0wsjeloKoMr6iCY6dP92H6Vw/oTyICIthibxjm/DfN9lVz8IqtqKYLUXfoKVMVQVVJOElGjrnnUt9T9wbgp8AyYKaGlqingHZU/uG2NTZSVqwHQTWkx9hxjkpWDaCg6Ckj5qebgBVbT3V3NNXMSiWSDdGV3hrtzla7J+duwPOToIg42ChPQOQjspnSlp1V+Gjdged7+8UN5CRAV7a5EdFNwCjEaBR27b3W890TE7g24NAP/mMDXRWrGoFPQI9ls/MWO2dWFAar/xcOIImbbpA3zgAAAABJRU5ErkJggg==);
```
[the remark this answers is not in the document]
Comment (mattpap) [reply bokehjs/src/less/logo.less:15]: Sure. This can be done, assuming appropriate images are available.

**Comment by bryevdv** [comment]
LGTM

**Pull request merged**